initial cypress setup and added home header test

ShaileshParmar11 opened the pull request:
> First thing, PLEASE READ THIS: [ReactPlay Code Review Checklist](https://github.com/reactplay/react-play/wiki/ReactPlay-Code-Review-Checklist)

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. 

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

ShaileshParmar11 commented:
PR is ready for review. 🚀

Sachin-chaurasiya reviewed (dismissed):
LGTM

atapas reviewed (commented):
@ShaileshParmar11 thanks for putting it together. Can we also make readme changes suggesting how to run the tests?

Also, can we make it run part of out build process and get a report?

atapas commented:
> > @ShaileshParmar11 thanks for putting it together. Can we also make readme changes suggesting how to run the tests?
> > Also, can we make it run part of out build process and get a report?
> 
> Like Github actions right for every PR?

Yep

ShaileshParmar11 commented:
> > > @ShaileshParmar11 thanks for putting it together. Can we also make readme changes suggesting how to run the tests?
> > > Also, can we make it run part of out build process and get a report?
> > 
> > 
> > Like Github actions right for every PR?
> 
> Yep

After lots of failed [attempts](https://github.com/ShaileshParmar11/react-play/actions) I am able to add github workflow with success, you might need to give permission to run the same from repo setting -> Actions -> general -> `Workflow permissions` -> read and write permission -> save,

atapas commented:
> > > > @ShaileshParmar11 thanks for putting it together. Can we also make readme changes suggesting how to run the tests?
> > > > Also, can we make it run part of out build process and get a report?
> > > 
> > > 
> > > Like Github actions right for every PR?
> > 
> > 
> > Yep
> 
> After lots of failed [attempts](https://github.com/ShaileshParmar11/react-play/actions) I am able to add github workflow with success, you might need to give permission to run the same from repo setting -> Actions -> general -> `Workflow permissions` -> read and write permission -> save,

Done

ShaileshParmar11 commented:
> > > > > @ShaileshParmar11 thanks for putting it together. Can we also make readme changes suggesting how to run the tests?
> > > > > Also, can we make it run part of out build process and get a report?
> > > > 
> > > > 
> > > > Like Github actions right for every PR?
> > > 
> > > 
> > > Yep
> > 
> > 
> > After lots of failed [attempts](https://github.com/ShaileshParmar11/react-play/actions) I am able to add github workflow with success, you might need to give permission to run the same from repo setting -> Actions -> general -> `Workflow permissions` -> read and write permission -> save,
> 
> Done

It is strange workflow is triggering in my own repo action but not here! do we need to do anything specific to run here? or do we need a workflow file in main to trigger?

koustov commented:
Few concerns
1. Is it running inheadless mode? Bt default cypress runs in headless but the video that has been tagged shows ``headed` mode?
2. Where its expected to run? In this repo action. 

> > > > > > @ShaileshParmar11 thanks for putting it together. Can we also make readme changes suggesting how to run the tests?
> > > > > > Also, can we make it run part of out build process and get a report?
> > > > > 
> > > > > 
> > > > > Like Github actions right for every PR?
> > > > 
> > > > 
> > > > Yep
> > > 
> > > 
> > > After lots of failed [attempts](https://github.com/ShaileshParmar11/react-play/actions) I am able to add github workflow with success, you might need to give permission to run the same from repo setting -> Actions -> general -> `Workflow permissions` -> read and write permission -> save,
> > 
> > 
> > Done
> 
> It is strange workflow is triggering in my own repo action but not here! do we need to do anything specific to run here? or do we need a workflow file in main to trigger?

What is the trigger?

koustov commented on .gitignore:
```diff
@@ -7,6 +7,8 @@
 
 # testing
 /coverage
+/cypress/screenshots
+/cypress/videos
```
what are these for?

ShaileshParmar11 replied:
we need to ignore this path as it captures the screenshot and video of the test which runs in headless mode, and of course we don't want to include those in Github :)

koustov commented on package.json:
```diff
@@ -87,6 +89,7 @@
   },
   "devDependencies": {
     "autoprefixer": "^10.4.7",
+    "cypress": "^10.6.0",
```
Can this be a global package instead of bundling it with app?

ShaileshParmar11 replied:
We need cypress dependency, I don't think without this it will run!

koustov commented on README.md:
````diff
@@ -132,6 +132,24 @@ yarn build
+yarn cypress:open
+```
+
+It will open cypress dashboard, through which you need to select `E2E Testing`.
````
I would suggest defining a boundary about "who should run these tests" and "when should it be run"

ShaileshParmar11 replied:
Anyone can run this test locally we can not restrict anyone, and there is no particular scenario in which we need to run the test as we are going to integrate GitHub workflow, which will take care of running cypress in every PR

koustov commented on cypress/support/commands.ts:
```diff
@@ -0,0 +1,37 @@
+/// <reference types="cypress" />
```
This can be moved to some doc/md page

ShaileshParmar11 replied:
I didn't get this. can you elaborate? 
```
/// <reference types="cypress" />
```
and this is to let file know about, this is cypress file so it provides relevant suggestions

koustov commented on cypress/support/constant.ts:
```diff
@@ -0,0 +1,2 @@
+export const TWEET_COUNT = 10;
+export const CONTRIBUTORS_COUNT = 29;
```
Length wouldn't be a great idea to put in TC

ShaileshParmar11 replied:
Sure it is as we are checking if the expected elements are rendered or not, in the case of tweets we have hard coded the logic in the frontend side so we know we should be expecting the 10 tweets on the website, so I am checking the same here that all the requested tweets are available on the page or not!

koustov reviewed (changes requested):
Well I have a few concerns that might not be related to PR however definitely associated with the  life term of e2e
1. Is it running in the headless mode? I know cypress by default runs in headless however the video demonstrates a `headed` manner. Any configuration update?
2. Do we need to associate any VM in the future to run extensive tests?
3. What is the triggering point are we looking for?
4. What is the priority list of TCs to capture within e2e
    - I know this PR is to prepare the infrastructure. Then I would suggest, not to push any TC as of now, 
5. As the element selection is based on ID so it would work great with controlled code. What about an uncontrolled one? I mean for individual plays? How can we bring them into pictures?

ShaileshParmar11 commented:
> Few concerns
> 
> 1. Is it running inheadless mode? Bt default cypress runs in headless but the video that has been tagged shows ``headed` mode?
> 2. Where its expected to run? In this repo action.
> 
> > > > > > > @ShaileshParmar11 thanks for putting it together. Can we also make readme changes suggesting how to run the tests?
> > > > > > > Also, can we make it run part of out build process and get a report?
> > > > > > 
> > > > > > 
> > > > > > Like Github actions right for every PR?
> > > > > 
> > > > > 
> > > > > Yep
> > > > 
> > > > 
> > > > After lots of failed [attempts](https://github.com/ShaileshParmar11/react-play/actions) I am able to add github workflow with success, you might need to give permission to run the same from repo setting -> Actions -> general -> `Workflow permissions` -> read and write permission -> save,
> > > 
> > > 
> > > Done
> > 
> > 
> > It is strange workflow is triggering in my own repo action but not here! do we need to do anything specific to run here? or do we need a workflow file in main to trigger?
> 
> What is the trigger?

can you check -> [here](https://github.com/ShaileshParmar11/react-play/actions), you will get more idea how cypress is run in CLI, also triggering means GitHub action didn't start as expected like it is started while I was testing this PR with my own branch

ShaileshParmar11 commented:
> Well I have a few concerns that might not be related to PR however definitely associated with the life term of e2e
> 
> 1. Is it running in the headless mode? I know cypress by default runs in headless however the video demonstrates a `headed` manner. Any configuration update?
video is just to demonstrate all test are passing and working fine, and yes this test will run in headless mode on CLI
> 2. Do we need to associate any VM in the future to run extensive tests?
Github actions will take's care of everything
> 3. What is the triggering point are we looking for?
I have added github workflow which is not running here so that what i am talking about!
> 4. What is the priority list of TCs to capture within e2e
>    I have added those cases which is added in project's wiki please refer [here](https://github.com/reactplay/react-play/wiki/End-to-End-Test-Scenarios)
>    * I know this PR is to prepare the infrastructure. Then I would suggest, not to push any TC as of now,
> 5. As the element selection is based on ID so it would work great with controlled code. What about an uncontrolled one? I mean for individual plays? How can we bring them into pictures?
you can write tests based on uncontrolled code as well but it would be preferable to use id to select any element because generally id will not change, but classes do

ShaileshParmar11 commented:
PR is Ready for Review

koustov commented:
@ShaileshParmar11 I don't see any workflow configured on this repo. Can you configure that before we merge it? You can bypass the failure until this PR is through

koustov commented:
My view
> Well I have a few concerns that might not be related to PR however definitely associated with the life term of e2e
> 
1. Is it running in the headless mode? I know cypress by default runs in headless however the video demonstrates a `headed` manner. Any configuration update?
video is just to demonstrate all test are passing and working fine, and yes this test will run in headless mode on CLI
**KM**: OK

2. Do we need to associate any VM in the future to run extensive tests?
Github actions will take's care of everything
**KM**: I understand however my point was, as the application is growing and the transactions are more request, does it make sense to offload the action (from workflow) to some VM? Anyway, I don't think we need to over-engineer it right now, so I taking this comment back

3. What is the triggering point are we looking for?
I have added github workflow which is not running here so that what i am talking about!
**KM**: I mean, what will be the action the workflow will depend on . Like push, commit, merge etc

4. What is the priority list of TCs to capture within e2e
   I have added those cases which is added in project's wiki please refer [here](https://github.com/reactplay/react-play/wiki/End-to-End-Test-Scenarios)
**KM**: OK
   * I know this PR is to prepare the infrastructure. Then I would suggest, not to push any TC as of now,
**KM**:  I am still inclining towards "not to check in any TC" for now. Just the infrastructure.

 5. As the element selection is based on ID so it would work great with controlled code. What about an uncontrolled one? I mean for individual plays? How can we bring them into pictures?
you can write tests based on uncontrolled code as well but it would be preferable to use id to select any element because generally id will not change, but classes do 
**KM**: We need to plan for a V2.0 of this where we would discuss. I have certain thought around it.

koustov reviewed (approved):
lgtm

atapas commented:
@all-contributors please add @ShaileshParmar11  for Code
